Implement growing logic for `TextAtlas`

hecrj opened the pull request:
The `TextAtlas` will have an initial size of `256` (we could make this configurable) and `try_allocate` will keep track of the glyphs in use in the current frame, returning `None` when the LRU glyph is in use.

In that case, `TextRenderer::prepare` will return
`PrepareError::AtlasFull` with the `ContentType` of the atlas that is full. The user of the library can then call `TextAtlas::grow` with the provided `ContentType` to obtain a bigger atlas (by `256`).

A `TextAtlas::grow` call clears the whole atlas and, as a result, all of the `prepare` calls need to be repeated in a frame until they all succeed. Overall, the atlas will rarely need to grow and so the calls will not need to be repated often.

Finally, the user needs to call `TextAtlas::trim` at the end of the frame. This allows us to clear the glyphs in use collection in the atlas. Maybe there is a better way to model this in an API that forces the user to trim the atlas (e.g. make `trim` return a new type and changing `prepare` and `render` to take that type instead).

genusistimelord commented:
I decided for my own Texture Storage to use Texture Layers With a Garbage Collector via 2 hash_maps As this made more Sense than LRU since I can trade between the 2 hash maps upon usage of the Glyph.

Basically you have an Empty HashMap and a Storage HashMap . When you use Get you set the Storage (key, Allocation) to the Empty hashmap, If its not in the Storage then we check the Empty if in neither return None. At the end of the frame you then check whats left inside the Storage Hashmap and clear it if we reached a Certain PressureMin and also having a PressureMax to prevent Further TextureAtlas layers. We then swap the keys to the Storage again when done and do the process again the next frame. 

If you Like this system instead I can implement it into Glyphon. As I see no need to keep a HashSet for used and a LRU cache that's unbounded as they should not be used unbounded. Plus LRU doesn't Guarantee everything is being used or not.  At least with the Above system we know for Fact what has been used or not within the last Frame. And we avoid clearing Data out that shouldn't be cleared. Instead in these cases this should be a hard error if there is no space left and we cant make any more space due to usage.

I base the Pressure upon Layers this way you can easily keep track of limits better as its hard to keep track of how many allocations can be possible within an Texture Atlas. due to the possibility of different sizes etc.

```rust

///contained within the image atlas
pub cache: [HashMap<U, Allocation<Data>>; 2],
/// When Eviction starts each Cleanup Call. This is the amount of layers.
/// 0 means it will never evict anything.
pub pressure_min: usize,
/// When the System will Error if reached. This is the max allowed Layers
/// Default is 256 as Most GPU allow a max of 256.
pub pressure_max: usize,

//get the Allocation from the old or the new
 pub fn get(&mut self, key: &U) -> Option<Allocation<Data>> {
        let alloc = self.cache[0].remove(key);

        if let Some(allocation) = alloc {
            self.cache[1].insert(key.clone(), allocation);
            Some(allocation)
        } else {
            self.cache[1].get(key).copied()
        }
    }
    
    
     pub fn clean(&mut self) {
        if self.layers.len() >= self.pressure_min {
            self.cache.swap(0, 1);

            for (_key, allocation) in self.cache[1].drain() {
                self.layers
                    .get_mut(allocation.layer)
                    .unwrap()
                    .allocator
                    .deallocate(allocation.allocation);
            }

            return;
        } else if self.cache[0].len() < self.cache[1].len() {
            self.cache.swap(0, 1);
        }

        let mut data = Vec::with_capacity(self.cache[1].len());

        for (key, allocation) in self.cache[1].drain() {
            data.push((key, allocation));
        }

        for (key, allocation) in data {
            self.cache[0].insert(key, allocation);
        }
    }
```

hecrj commented:
@genusistimelord I don't understand.

- The purpose of the LRU is to evict the oldest glyphs first when strictly necessary. There is no need to deallocate every unused glyph every frame.
- We keep track of the glyphs used with the `glyphs_in_use` set. There is no need for 2 maps.
- Unbounded shouldn't be a problem since the atlas itself has a defined bound (its size).

genusistimelord commented:
Ok yeah After discussing this with you I agree but I do think we need Texture Atlas Layering to avoid any pitfalls as resizing a Texture Atlas to a new size really doesn't solves the issue and requires use to re-upload all the glyphs currently being used individually. So i would think it be more Efficient in the long run to have the ability to use Layers and make it so a new layer is only created when there is no room in the Atlas and no room can be made by de-allocating unused Glyphs.

grovesNL commented:
Good point, opened https://github.com/nical/etagere/issues/22

We also might be able to work around it in the meantime by internally remapping all of the old UVs to the new UVs, but that could be complicated (especially for clipped glyphs) without storing extra info during `prepare`.

hecrj commented:
I have rebased on top of `main` and also implemented automatic re-upload of the glyphs when growing the atlas in 65276a15bdf138da3f89cc82906c3775f1e0e7b9, which relies on https://github.com/nical/etagere/pull/24.

grovesNL commented on src/text_atlas.rs:
```diff
@@ -68,20 +76,121 @@ impl InnerAtlas {
+                (key, value) = self.glyph_cache.peek_lru()?;
+            }
+
+            if self.glyphs_in_use.contains(&key) {
```
Do we need to check `self.glyphs_in_use` for each of the keys in the loop too?

hecrj replied:
Yes, you are right. Given we can `grow` now, we should never `pop` any glyphs that are in use. Fixed in b979122dfa80e2072f0948aa227712201e173cc4.

grovesNL commented on src/text_atlas.rs:
```diff
@@ -68,20 +76,121 @@ impl InnerAtlas {
+        }
+
+        // TODO: Better resizing logic (?)
+        let new_size = (self.size + Self::INITIAL_SIZE).min(self.max_texture_dimension_2d);
```
A growth factor might be nice to avoid a lot of small allocations if the initial size is small, but we could add that later. e.g., maybe defaulting to 1.5x or 2x

hecrj replied:
Yeah. Let's add that later. The new `grow` operation in `etagere` works well with multiples of the initial size.

grovesNL reviewed (commented):
Thank you!

hecrj commented:
https://github.com/nical/etagere/pull/24 was finally merged, which should unblock this! I have updated `etagere` and hopefully fixed any outstanding issues!

grovesNL reviewed (approved):
Looks great, thank you!
